catalyst_voices/lib/widgets/common/tab_bar_stack_view.dart

```diff
@@ -0,0 +1,144 @@
+import 'package:collection/collection.dart';
+import 'package:flutter/material.dart';
+
+/// This widget is every similar to [TabBarView], and is meant to be used
+/// together with [TabBar], but displays children in non scrollable faction
+/// thanks to [IndexedStack].
+class TabBarStackView extends StatefulWidget {
+  /// This widget's selection and animation state.
+  ///
+  /// If [TabController] is not provided, then the value of
+  /// [DefaultTabController.of] will be used.
+  final TabController? controller;
+
+  /// One widget per tab.
+  ///
+  /// Its length must match the length of the [TabBar.tabs]
+  /// list, as well as the [controller]'s [TabController.length].
+  final List<Widget> children;
+
+  const TabBarStackView({
+    super.key,
+    this.controller,
+    required this.children,
+  });
+
+  @override
+  State<TabBarStackView> createState() => _TabBarStackViewState();
+}
+
+class _TabBarStackViewState extends State<TabBarStackView> {
+  TabController? _controller;
+
+  int? _currentIndex;
+
+  // If the TabBarView is rebuilt with a new tab controller, the caller should
+  // dispose the old one. In that case the old controller's animation will be
+  // null and should not be accessed.
+  bool get _controllerIsValid => _controller?.animation != null;
+
+  @override
+  void didChangeDependencies() {
+    super.didChangeDependencies();
+    _updateTabController();
+    _currentIndex = _controller!.index;
+  }
+
+  @override
+  void didUpdateWidget(TabBarStackView oldWidget) {
+    super.didUpdateWidget(oldWidget);
+    if (widget.controller != oldWidget.controller) {
+      _updateTabController();
+      _currentIndex = _controller!.index;
+    }
+  }
+
+  @override
+  void dispose() {
+    if (_controllerIsValid) {
+      _controller!.animation!.removeListener(_handleTabControllerAnimationTick);
+    }
+    _controller = null;
+    // We don't own the _controller Animation, so it's not disposed here.
+    super.dispose();
+  }
+
+  @override
+  Widget build(BuildContext context) {
+    return _TabsBodyContainer(
+      currentIndex: _currentIndex,
+      children: widget.children,
+    );
+  }
+
+  void _updateTabController() {
+    final newController =
+        widget.controller ?? DefaultTabController.maybeOf(context);
+
+    assert(
+      newController != null,
+      'No TabController for ${widget.runtimeType}.\n'
+      'When creating a ${widget.runtimeType}, you must either provide an '
+      'explicit '
+      'TabController using the "controller" property, or you must ensure that '
+      'there '
+      'is a DefaultTabController above the ${widget.runtimeType}.\n'
+      'In this case, there was neither an explicit controller nor a default '
+      'controller.',
+    );
+
+    if (newController == _controller) {
+      return;
+    }
+
+    if (_controllerIsValid) {
+      _controller!.animation!.removeListener(_handleTabControllerAnimationTick);
```

Comment:
would it make sense to rewrite this a bit to distinguish on which controller you add / remove listener? I.e.:
```dart
  final oldController = _controller;
  final newController = ...;
  
  oldController?.animation?.removeListener(_handleTabControllerAnimationTick);
  newController!.animation!.addListener(_handleTabControllerAnimationTick);
```

using the same variable on which you both remove and add listener is a bit tricky to understand